### ISSUE TITLE
t/op/coreamp.t:  revisions in 'evalbytes' block flawed

### ISSUE THREAD

**Issue by jkeenan** [atoomic/perl#268]
As originally noted in https://github.com/atoomic/perl/issues/267#issuecomment-689743392, revisions that were made to the `evalbytes` block in `t/op/coreamp.t` are fundamentally flawed regardless of the locale problem which was the focus of that ticket.

Currently, in the `alpha-dev-02-strict` branch, we have this code:
```
 517 {
 518   my $U_100_bytes = byte_utf8a_to_utf8n("\xc4\x80");
 519   chop(my $upgraded = "use utf8; { no strict; $U_100_bytes" . chr 256);
 520   $upgraded .= " }";
 521   is( &myevalbytes($upgraded), chr 256, '&evalbytes [myevalbytes]' );
 522   # Test hints
 523   require strict;
 524   strict->import; # cannot use strict to check hints
 525   my $hint_bit = 0x20000; # use an arbitrary bit
 526   $^H |= $hint_bit;
 527   my $x = &myevalbytes('my $x; BEGIN { $x = $^H } $x');
 528   is( $x & $hint_bit, 0, "run-time hint bits do not leak into &evalbytes" )     ;
 529   $^H |= ~$hint_bit; # remove the bit
```
At line 525, the value assigned to `$hint_bit` is `0x20000`, which is described as 'arbitrary'.  When I run the file, in this section of the output I get:
```
ok 96 - &evalbytes with too many args ; line:75
ok 97 - &evalbytes [myevalbytes]
ok 98 - run-time hint bits do not leak into &evalbytes
ok 99 - compile-time hh propagates into &evalbytes
ok 100 - compile-time hint bits propagate into &evalbytes
```
Now suppose that I simply assign a different value to `$hint_bit`, say, '0x1ffff'.
```
$ git diff
diff --git a/t/op/coreamp.t b/t/op/coreamp.t
index cb413a13ac..824cb55af8 100644
--- a/t/op/coreamp.t
+++ b/t/op/coreamp.t
@@ -522,7 +522,7 @@ $main::tests += 4;
   # Test hints
   require strict;
   strict->import; # cannot use strict to check hints
-  my $hint_bit = 0x20000; # use an arbitrary bit
+  my $hint_bit = 0x1ffff; # use an arbitrary bit
   $^H |= $hint_bit;
   my $x = &myevalbytes('my $x; BEGIN { $x = $^H } $x');
   is( $x & $hint_bit, 0, "run-time hint bits do not leak into &evalbytes" );
```
Now, when I run the file, I get:
```
$ cd t; ./perl harness -v op/coreamp.t ; cd -
...
ok 96 - &evalbytes with too many args ; line:75
ok 97 - &evalbytes [myevalbytes]
not ok 98 - run-time hint bits do not leak into &evalbytes
ok 99 - compile-time hh propagates into &evalbytes
ok 100 - compile-time hint bits propagate into &evalbytes
ok 101 - &exit with too many args ; line:97
# Failed test 98 - run-time hint bits do not leak into &evalbytes at op/coreamp.t line 528
#      got "1538"
# expected "0"
```
So, even before we toss locales into the mix, it seems to me that test 98 in `t/op/coreamp.t` is flawed.

These tests need to be re-thought.

**Comment by jkeenan** [comment]
> 
> So, even before we toss locales into the mix, it seems to me that test 98 in `t/op/coreamp.t` is flawed.
> 
> These tests need to be re-thought.

I decided to go back to the way the tests were written at the point where we forked from Perl 5 blead.
```
$ git diff |cat
diff --git a/t/op/coreamp.t b/t/op/coreamp.t
index cb413a13ac..0a819e5a1f 100644
--- a/t/op/coreamp.t
+++ b/t/op/coreamp.t
@@ -516,18 +516,14 @@ test_proto 'evalbytes';
 $main::tests += 4;
 {
   my $U_100_bytes = byte_utf8a_to_utf8n("\xc4\x80");
-  chop(my $upgraded = "use utf8; { no strict; $U_100_bytes" . chr 256);
-  $upgraded .= " }";
-  is( &myevalbytes($upgraded), chr 256, '&evalbytes [myevalbytes]' );
+  chop(my $upgraded = "use utf8; $U_100_bytes" . chr 256);
+  is &myevalbytes($upgraded), chr 256, '&evalbytes';
   # Test hints
   require strict;
-  strict->import; # cannot use strict to check hints
-  my $hint_bit = 0x20000; # use an arbitrary bit
-  $^H |= $hint_bit;
-  my $x = &myevalbytes('my $x; BEGIN { $x = $^H } $x');
-  is( $x & $hint_bit, 0, "run-time hint bits do not leak into &evalbytes" );
-  $^H |= ~$hint_bit; # remove the bit
-
+  strict->import;
+  &myevalbytes('
+    is someone, "someone", "run-time hint bits do not leak into &evalbytes"
+  ');
   use strict;
   BEGIN { $^H{coreamp} = 42 }
   $^H{coreamp} = 75;
```
Running the test file at this point gives 1 test failure and -- curiously -- one test not run.
```
$ cd t; ./perl harness -v op/coreamp.t  1>../err 2>&1; cd -; vi err
# Failed test 97 - &evalbytes at op/coreamp.t line 520
#      got undef
# expected "\x{100}"
# Looks like you planned 777 tests but ran 776.
op/coreamp.t ..
ok 1 - no crash with &CORE::foo when *_{ARRAY} is undef
...
ok 96 - &evalbytes with too many args ; line:75
not ok 97 - &evalbytes
ok 98 - compile-time hh propagates into &evalbytes
ok 99 - compile-time hint bits propagate into &evalbytes
```
Where did this test go to?
```
  &myevalbytes('
    is someone, "someone", "run-time hint bits do not leak into &evalbytes"
  ');
```
I then commented out all statements that mentioned `strict` because they're irrelevant once we're in a strict-by-default environment.
```
diff --git a/t/op/coreamp.t b/t/op/coreamp.t
index 0a819e5a1f..e8727eb6d5 100644
--- a/t/op/coreamp.t
+++ b/t/op/coreamp.t
@@ -519,12 +519,12 @@ $main::tests += 4;
   chop(my $upgraded = "use utf8; $U_100_bytes" . chr 256);
   is &myevalbytes($upgraded), chr 256, '&evalbytes';
   # Test hints
-  require strict;
-  strict->import;
+#  require strict;
+#  strict->import;
   &myevalbytes('
     is someone, "someone", "run-time hint bits do not leak into &evalbytes"
   ');
-  use strict;
+#  use strict;
   BEGIN { $^H{coreamp} = 42 }
   $^H{coreamp} = 75;
   &myevalbytes('
```
I got the same test results as above.

I then added debugging statements:
```
diff --git a/t/op/coreamp.t b/t/op/coreamp.t
index e8727eb6d5..e006ae216d 100644
--- a/t/op/coreamp.t
+++ b/t/op/coreamp.t
@@ -517,14 +517,15 @@ $main::tests += 4;
 {
   my $U_100_bytes = byte_utf8a_to_utf8n("\xc4\x80");
   chop(my $upgraded = "use utf8; $U_100_bytes" . chr 256);
+print STDERR "AAA: \$U_100_bytes: $U_100_bytes\n";
+print STDERR "BBB: \$upgraded:    $upgraded\n";
+print STDERR "CCC: ", &myevalbytes($upgraded), "\n";
+print STDERR "DDD: ", CORE::evalbytes($upgraded), "\n";
   is &myevalbytes($upgraded), chr 256, '&evalbytes';
   # Test hints
-#  require strict;
-#  strict->import;
   &myevalbytes('
     is someone, "someone", "run-time hint bits do not leak into &evalbytes"
   ');
-#  use strict;
   BEGIN { $^H{coreamp} = 42 }
   $^H{coreamp} = 75;
   &myevalbytes('
```
Running the file:
```
AAA: $U_100_bytes: Ā
BBB: $upgraded:    use utf8; Ā
CCC:
DDD:
# Failed test 97 - &evalbytes at op/coreamp.t line 524
#      got undef
# expected "\x{100}"
# Looks like you planned 777 tests but ran 776.
op/coreamp.t ..
ok 1 - no crash with &CORE::foo when *_{ARRAY} is undef
...
ok 96 - &evalbytes with too many args ; line:75
not ok 97 - &evalbytes
ok 98 - compile-time hh propagates into &evalbytes
ok 99 - compile-time hint bits propagate into &evalbytes
```
Still no "someone" test -- but what's up with those blanks at CCC and DDD?  Time to RTFM!
```
$ perldoc -f evalbytes
...
If called when "use utf8" is in effect, the string will be
assumed to be encoded in UTF-8, and "evalbytes" will make a
temporary copy to work from, downgraded to non-UTF-8. If this is
not possible (because one or more characters in it require
UTF-8), the "evalbytes" will fail with the error stored in $@.
```
Hmm, maybe those blanks appear when `&myevalbytes` fails?  So let's check for errors.
```
diff --git a/t/op/coreamp.t b/t/op/coreamp.t
index e006ae216d..76a5a5c59d 100644
--- a/t/op/coreamp.t
+++ b/t/op/coreamp.t
@@ -521,11 +521,19 @@ print STDERR "AAA: \$U_100_bytes: $U_100_bytes\n";
 print STDERR "BBB: \$upgraded:    $upgraded\n";
 print STDERR "CCC: ", &myevalbytes($upgraded), "\n";
 print STDERR "DDD: ", CORE::evalbytes($upgraded), "\n";
-  is &myevalbytes($upgraded), chr 256, '&evalbytes';
+  {
+    local $@;
+    is &myevalbytes($upgraded), chr 256, '&evalbytes';
+    print STDERR "EEE: <$@>\n";
+  }
   # Test hints
-  &myevalbytes('
-    is someone, "someone", "run-time hint bits do not leak into &evalbytes"
-  ');
+  {
+    local $@;
+    &myevalbytes('
+      is someone, "someone", "run-time hint bits do not leak into &evalbytes"
+    ');
+    print STDERR "FFF: <$@>\n";
+  }
   BEGIN { $^H{coreamp} = 42 }
   $^H{coreamp} = 75;
   &myevalbytes('
```
Running the file:
```
AAA: $U_100_bytes: Ā
BBB: $upgraded:    use utf8; Ā
CCC:
DDD:
# Failed test 97 - &evalbytes at op/coreamp.t line 526
#      got undef
# expected "\x{100}"
EEE: <Bareword "Ā" not allowed while "strict subs" in use at (eval 69)[op/coreamp.t:526] line 1.
>
FFF: <Bareword "someone" not allowed while "strict subs" in use at (eval 70)[op/coreamp.t:532] line 2.
>
# Looks like you planned 777 tests but ran 776.
op/coreamp.t ..
```
*Thar she blows, mateys!*  Once we check for errors, we find locations where we have to relax strictures.  Doing so, while removing debugging code.
```
diff --git a/t/op/coreamp.t b/t/op/coreamp.t
index 76a5a5c59d..7b4216767d 100644
--- a/t/op/coreamp.t
+++ b/t/op/coreamp.t
@@ -517,22 +517,12 @@ $main::tests += 4;
 {
   my $U_100_bytes = byte_utf8a_to_utf8n("\xc4\x80");
   chop(my $upgraded = "use utf8; $U_100_bytes" . chr 256);
-print STDERR "AAA: \$U_100_bytes: $U_100_bytes\n";
-print STDERR "BBB: \$upgraded:    $upgraded\n";
-print STDERR "CCC: ", &myevalbytes($upgraded), "\n";
-print STDERR "DDD: ", CORE::evalbytes($upgraded), "\n";
   {
-    local $@;
+    no strict 'subs';
     is &myevalbytes($upgraded), chr 256, '&evalbytes';
-    print STDERR "EEE: <$@>\n";
-  }
-  # Test hints
-  {
-    local $@;
     &myevalbytes('
       is someone, "someone", "run-time hint bits do not leak into &evalbytes"
     ');
-    print STDERR "FFF: <$@>\n";
   }
   BEGIN { $^H{coreamp} = 42 }
   $^H{coreamp} = 75;
```
Running the file:
```
$ cd t; ./perl harness -v op/coreamp.t  1>../err 2>&1; cd -; vi err
...
ok 96 - &evalbytes with too many args ; line:75
ok 97 - &evalbytes
ok 98 - run-time hint bits do not leak into &evalbytes
ok 99 - compile-time hh propagates into &evalbytes
ok 100 - compile-time hint bits propagate into &evalbytes
```
The `run-time hint bits do not leak into &evalbytes` test is now being run and all tests are passing.  Moreover, all tests also pass when locales are thown into the mix.
```
$ cd t; PERL_IO= LC_ALL=en_US.utf8 PERL_UNICODE= ./perl harness -v op/coreamp.t  1>../err 2>&1; cd -; vi err
...
op/coreamp.t ..
ok 1 - no crash with &CORE::foo when *_{ARRAY} is undef
...
ok 96 - &evalbytes with too many args ; line:75
ok 97 - &evalbytes
ok 98 - run-time hint bits do not leak into &evalbytes
ok 99 - compile-time hh propagates into &evalbytes
ok 100 - compile-time hint bits propagate into &evalbytes
...
ok 777 - no implicit $_ for second argument to mkdir
1..777
ok
All tests successful.
Files=1, Tests=777,  1 wallclock secs ( 0.05 usr  0.00 sys +  0.10 cusr  0.00 csys =  0.15 CPU)
Result: PASS
```

**Comment by jkeenan** [comment]
Fixed in commit 641918666c.

Thank you very much.
Jim Keenan
